libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1334,28 +1335,111 @@ string YulUtilFunctions::allocationFunction()
+	solAssert(_type.baseType()->hasSimpleZeroValueInMemory(), "");
 
-	string functionName = "allocate_memory_array_" + _type.identifier();
+	string functionName = "zero_simple_memory_array_" + _type.identifier();
 	return m_functionCollector.createFunction(functionName, [&]() {
 		return Whiskers(R"(
-			function <functionName>(length) -> memPtr {
-				memPtr := <alloc>(<allocSize>(length))
+			function <functionName>(memPtr, allocSize) {
+				calldatacopy(memPtr, calldatasize(), allocSize)
+			}
+		)")
+		("functionName", functionName)
+		.render();
+	});
+}
+
+string YulUtilFunctions::zeroComplexMemoryArrayFunction(ArrayType const& _type)
+{
+	solAssert(!_type.baseType()->hasSimpleZeroValueInMemory(), "");
+
+	string functionName = "zero_complex_memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(memPtr, allocSize) {
+				let firstSlot := memPtr
+				let dataSize := allocSize
 				<?dynamic>
-				mstore(memPtr, length)
+				firstSlot := add(firstSlot, 32)
+				dataSize := sub(dataSize, 32)
 				</dynamic>
+				let words := div(add(dataSize, 31), 32)
+				for {let i := 0} lt(i, dataSize) { i := add(i, 32) } {
+					mstore(add(firstSlot, i), <zeroValue>())
+				}
 			}
 		)")
 		("functionName", functionName)
-		("alloc", allocationFunction())
-		("allocSize", arrayAllocationSizeFunction(_type))
 		("dynamic", _type.isDynamicallySized())
+		("zeroValue", zeroValueFunction(*_type.baseType()))
 		.render();
 	});
 }
 
+string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type, bool _shouldZero)
+{
+	solUnimplementedAssert(!_type.isByteArray(), "");
+
+	string functionName = "allocate_" + string(_shouldZero ? "and_zero_" : "") + "memory_array_" + _type.identifier();
+	return m_functionCollector.createFunction(functionName, [&]() {
+		auto const* base = _type.baseType();
+		solUnimplementedAssert(base->sizeOnStack() == 1, "Stacksize not yet implemented!");
+		return Whiskers(R"(
+				function <functionName>(length) -> memPtr {
+					let allocSize := <allocSize>(length)
+					memPtr := <alloc>(allocSize)
+					<?shouldZero>
+						<zeroArrayFunction>(memPtr, allocSize)
+					</shouldZero>
+					<?dynamic>
+					mstore(memPtr, length)
+					</dynamic>
+				}
+			)")
+			("functionName", functionName)
+			("alloc", allocationFunction())
+			("allocSize", arrayAllocationSizeFunction(_type))
+			("shouldZero", _shouldZero)
+			("zeroArrayFunction", base->hasSimpleZeroValueInMemory() ? zeroSimpleMemoryArrayFunction(_type) : zeroComplexMemoryArrayFunction(_type))
+			("dynamic", _type.isDynamicallySized())
+			.render();
+	});
+}
+
+string YulUtilFunctions::allocateMemoryStructFunction(StructType const& _type, bool _shouldZero)
+{
+	string functionName = "allocate_" + string(_shouldZero ? "and_zero_" : "") + "memory_struct_" + _type.identifier();
+	Whiskers templ(R"(
+	function <functionName>() -> memPtr {
+		memPtr := <alloc>(<memberSize>)
```

Comment:
Don't you need to mulitply this by 32?

Reply:
And also please rename this to `memberCount`

Reply:
Can also multiply by 32 outside if combined with the assert suggested